subsys/bluetooth/host/adv.c

```diff
@@ -524,29 +535,29 @@ static int hci_set_ad(uint16_t hci_op, const struct bt_ad *ad, size_t ad_len)
 }
 
 static int hci_set_adv_ext_complete(struct bt_le_ext_adv *adv, uint16_t hci_op,
-				    const struct bt_ad *ad, size_t ad_len)
+				    size_t total_data_len, const struct bt_ad *ad, size_t ad_len)
 {
 	struct bt_hci_cp_le_set_ext_adv_data *set_data;
 	struct net_buf *buf;
 	int err;
-	uint8_t max_data_size;
+	const size_t cmd_size = total_data_len +
+				offsetof(struct bt_hci_cp_le_set_ext_adv_data, data);
 
-	buf = bt_hci_cmd_create(hci_op, sizeof(*set_data));
+	if (!atomic_test_bit(adv->flags, BT_ADV_EXT_ADV) &&
+	    total_data_len > BT_GAP_ADV_MAX_ADV_DATA_LEN) {
+		return -EINVAL;
+	}
+
+	buf = bt_hci_cmd_create(hci_op, cmd_size);
 	if (!buf) {
 		return -ENOBUFS;
 	}
 
-	set_data = net_buf_add(buf, sizeof(*set_data));
-	(void)memset(set_data, 0, sizeof(*set_data));
+	set_data = net_buf_add(buf, cmd_size);
+	(void)memset(set_data, 0, cmd_size);
 
-	if (atomic_test_bit(adv->flags, BT_ADV_EXT_ADV)) {
-		max_data_size = BT_HCI_LE_EXT_ADV_FRAG_MAX_LEN;
-	} else {
-		max_data_size = BT_GAP_ADV_MAX_ADV_DATA_LEN;
-	}
-
-	err = set_data_add_complete(set_data->data, max_data_size, ad, ad_len,
-				    &set_data->len);
+	err = set_data_add_complete(set_data->data, total_data_len,
+				    ad, ad_len, &set_data->len);
```

Comment:
Correct me if I am wrong, but won't `total_data_len` and `set_data->len` always be the same value after this call?

Reply:
Yes, always same, so we can use `(void)set_data_add_complete`.

Good point.

Reply:
> Yes, always same, so we can use `(void)set_data_add_complete`.
> 
> Good point.

How did you come to that conclusion? 

It does actually looks like that in the case of `BT_DATA_NAME_COMPLETE` that is shortened that they will not be equal. In that case, the calculated length is larger than &set_data->len. 

The case where shortening the name fails due to the size can still happen here, can't it?

Reply:
`if ((set_data_len + len + 2) > set_data_len_max) `

This judgment will also false, because we do same calculation for set_data_len_max before. Only problem is we can't change type to short_name，or we not need todo this, i think.

Reply:
``` C
	/* Provide the opportunity to truncate the complete name */
	if (!atomic_test_bit(adv->flags, BT_ADV_EXT_ADV) &&
	    total_data_len > BT_GAP_ADV_MAX_ADV_DATA_LEN) {
		total_data_len = BT_GAP_ADV_MAX_ADV_DATA_LEN;
	}
```

Solved.